Cabal/doc/nix-local-build.markdown

```diff
@@ -0,0 +1,1168 @@
+    # (Note: By default, if we have a local package we will
+    # automatically use it, so it generally not be necessary to
+    # specify this)
+    constraints: bar source
+
+    # Require that bar be solved with test suites and benchmarks enabled
+    # (Note: By default, new-build configures the solver to make
+    # a best-effort attempt to enable these stanzas, so this generally
+    # should not be necessary.)
+    constraints: bar test,
+                 bar bench
+    ~~~~~~~~~~~~~~~~
+
+    The command line variant of this field is `--constraint="pkg >= 2.0"`; to specify
+    multiple constraints, pass the flag multiple times.
+
+`preferences:` _preference_ (comma separated)
+:   Like `constraints`, but the solver will attempt to satisfy these
+    preferences on a best-effort basis.  The resulting install is
+    locally optimal with respect to preferences; specifically,
+    no single package could be replaced with a more preferred version
+    that still satisfies the hard constraints.
+
+    Operationally, preferences can cause the solver to attempt certain
+    version choices of a package before others, which can improve
+    dependency solver runtime.
+
+    One way to use `preferences` is to take a known working set of
+    constraints (e.g., via `cabal new-freeze`) and record them as
+    preferences. In this case, the solver will first attempt to use this
+    configuration, and if this violates hard constraints, it will try to
+    find the minimal number of upgrades to satisfy the hard constraints
+    again.
+
+    The command line variant of this field is `--preference="pkg >= 2.0"`; to specify
+    multiple preferences, pass the flag multiple times.
+
+`allow-newer:` `none` _or_ `all` _or_ _list of scoped package names_ (space or comma separated, default: `none`)
+:   Allow the solver to pick an older version of some packages than
+    would normally be permitted by than the `build-depends` bounds
+    of packages in the install plan.  This option may be useful if
+    the dependency solver cannot otherwise find a valid install plan.
+
+    For example, to relax `pkg`s `build-depends` upper bound on
+    `dep-pkg`, write a scoped package name of the form:
+
+    ~~~~~~~~~~~~~~~~
+    allow-newer: pkg:dep-pkg
+    ~~~~~~~~~~~~~~~~
+
+    This syntax is recommended, as it is often only a single package
+    whose upper bound is misbehaving.  In this case, the upper bounds of
+    other packages should still be respected; indeed, relaxing the bound
+    can break some packages which test the selected version of packages.
+
+    However, in some situations (e.g., when attempting to build
+    packages on a new version of GHC), it is useful to disregard
+    *all* upper-bounds, with respect to a package or all packages.
+    This can be done by specifying just a package name, or using
+    the keyword `all` to specify all packages:
+
+    ~~~~~~~~~~~~~~~~
+    # Disregard upper bounds involving the packages bar, baz and quux
```

Comment:
It isn't 100% clear which side of the dependency bar is on.  Something like "bounds involving dependencies on the packages bar ..." might be clearer.